src/rust/engine/fs/src/lib.rs

```diff
@@ -297,64 +381,131 @@ impl PathGlob {
+  // TODO(cosmicexplorer): match this up with the allowed values for the GlobMatchErrorBehavior type
+  // in python somehow?
+  pub fn create(behavior: &str) -> Result<Self, String> {
+    match behavior {
+      "ignore" => Ok(StrictGlobMatching::Ignore),
+      "warn" => Ok(StrictGlobMatching::Warn),
+      "error" => Ok(StrictGlobMatching::Error),
+      _ => Err(format!(
+        "Unrecognized strict glob matching behavior: {}.",
+        behavior,
+      )),
+    }
+  }
+
+  pub fn should_check_glob_matches(&self) -> bool {
+    match self {
+      &StrictGlobMatching::Ignore => false,
+      _ => true,
+    }
+  }
+
+  pub fn should_throw_on_error(&self) -> bool {
+    match self {
+      &StrictGlobMatching::Error => true,
+      _ => false,
+    }
+  }
+}
+
 #[derive(Debug)]
 pub struct PathGlobs {
-  include: Vec<PathGlob>,
-  exclude: Arc<Gitignore>,
+  include: Vec<PathGlobIncludeEntry>,
+  exclude: Arc<GitignoreStyleExcludes>,
+  strict_match_behavior: StrictGlobMatching,
 }
 
 impl PathGlobs {
-  pub fn create(include: &[String], exclude: &[String]) -> Result<PathGlobs, String> {
-    let ignore_for_exclude = if exclude.is_empty() {
-      EMPTY_IGNORE.clone()
-    } else {
-      Arc::new(create_ignore(exclude)
-        .map_err(|e| format!("Could not parse glob excludes {:?}: {:?}", exclude, e))?)
-    };
+  pub fn create(
+    include: &[String],
+    exclude: &[String],
+    strict_match_behavior: StrictGlobMatching,
+  ) -> Result<PathGlobs, String> {
+    let include = PathGlob::spread_filespecs(include)?;
+    Self::create_with_globs_and_match_behavior(include, exclude, strict_match_behavior)
+  }
+
+  fn create_with_globs_and_match_behavior(
+    include: Vec<PathGlobIncludeEntry>,
+    exclude: &[String],
+    strict_match_behavior: StrictGlobMatching,
+  ) -> Result<PathGlobs, String> {
+    let gitignore_excludes = GitignoreStyleExcludes::create(exclude)?;
     Ok(PathGlobs {
-      include: PathGlob::create(include)?,
-      exclude: ignore_for_exclude,
+      include,
+      exclude: gitignore_excludes,
+      strict_match_behavior,
     })
   }
 
-  pub fn from_globs(include: Vec<PathGlob>) -> PathGlobs {
-    PathGlobs {
-      include: include,
-      exclude: EMPTY_IGNORE.clone(),
-    }
+  pub fn from_globs(include: Vec<PathGlob>) -> Result<PathGlobs, String> {
+    let include = include
+      .into_iter()
+      .map(|glob| PathGlobIncludeEntry {
+        input: MISSING_GLOB_SOURCE.clone(),
+        globs: vec![glob],
+      })
+      .collect();
+    // An empty exclude becomes EMPTY_IGNORE.
+    PathGlobs::create_with_globs_and_match_behavior(include, &vec![], StrictGlobMatching::Ignore)
   }
 }
 
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub enum GlobSource {
+  ParsedInput(GlobParsedSource),
+  ParentGlob(PathGlob),
+}
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct GlobWithSource {
+  path_glob: PathGlob,
+  source: GlobSource,
+}
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
```

Comment:
For enums that don't own any data (or even enums that own very small amounts of data), it's generally fine to derive `Copy`, which avoids having to manually call `Clone`.